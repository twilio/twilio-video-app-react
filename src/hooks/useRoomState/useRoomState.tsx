import { useEffect, useState } from 'react';
import useVideoContext from '../useVideoContext/useVideoContext';

type RoomStateType = 'disconnected' | 'connected' | 'reconnecting';

export default function useRoomState() {
  const { room } = useVideoContext();
  const [state, setState] = useState<RoomStateType>('disconnected');

  useEffect(() => {
    if (room) {
      const setRoomState = () => setState(room.state as RoomStateType);
      setRoomState();
      room
        .on('disconnected', setRoomState)
        .on('reconnected', setRoomState)
        .on('reconnecting', setRoomState);
      return () => {
        room
          .off('disconnected', setRoomState)
          .off('reconnected', setRoomState)
          .off('reconnecting', setRoomState);
      };
    }
  }, [room]);

  return state;
}
