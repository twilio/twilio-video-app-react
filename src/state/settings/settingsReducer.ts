import { Track, VideoBandwidthProfileOptions } from 'twilio-video';

export interface Settings {
  trackSwitchOffMode: VideoBandwidthProfileOptions['trackSwitchOffMode'];
  dominantSpeakerPriority?: Track.Priority;
  bandwidthProfileMode: VideoBandwidthProfileOptions['mode'];
  maxAudioBitrate: string;
  contentPreferencesMode?: 'auto' | 'manual';
  clientTrackSwitchOffControl?: 'auto' | 'manual';
}

type SettingsKeys = keyof Settings;

export interface SettingsAction {
  name: SettingsKeys;
  value: string;
}

export const initialSettings: Settings = {
  trackSwitchOffMode: undefined,
  dominantSpeakerPriority: 'standard',
  bandwidthProfileMode: 'collaboration',
  maxAudioBitrate: '16000',
  contentPreferencesMode: 'auto',
  clientTrackSwitchOffControl: 'auto',
};

// This inputLabels object is used by ConnectionOptions.tsx. It is used to populate the id, name, and label props
// of the various input elements. Using a typed object like this (instead of strings) eliminates the possibility
// of there being a typo.
export const inputLabels = (() => {
  const target: any = {};
  for (const setting in initialSettings) {
    target[setting] = setting as SettingsKeys;
  }
  return target as { [key in SettingsKeys]: string };
})();

export function settingsReducer(state: Settings, action: SettingsAction) {
  return {
    ...state,
    [action.name]: action.value === 'default' ? undefined : action.value,
  };
}
