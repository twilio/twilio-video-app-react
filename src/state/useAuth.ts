import { useCallback, useEffect, useState } from 'react';
import * as firebase from 'firebase/app';
import 'firebase/auth';

const firebaseConfig = {
  apiKey: 'AIzaSyD3HQ2F0OO1oQgkHxnHoqLEtUhL7laJcAE',
  authDomain: 'video-app-79418.firebaseapp.com',
  databaseURL: 'https://video-app-79418.firebaseio.com',
  storageBucket: 'video-app-79418.appspot.com',
  messagingSenderId: '285008367772',
};

firebase.initializeApp(firebaseConfig);

const provider = new firebase.auth.GoogleAuthProvider();
provider.addScope('https://www.googleapis.com/auth/plus.login');

export default function useAuth(setToken: any) {
  const [user, setUser] = useState<firebase.User | null>(null);
  const [isAuthReady, setIsReady] = useState(false);

  useEffect(() => {
    firebase.auth().onAuthStateChanged(user => {
      setUser(user);
      setIsReady(true);
    });
  }, []);

  const signIn = useCallback(() => {
    return firebase
      .auth()
      .signInWithPopup(provider)
      .then(user => {
        setUser(user.user);
      });
  }, []);

  const signOut = useCallback(() => {
    return firebase
      .auth()
      .signOut()
      .then(() => {
        setUser(null);
        setToken('');
      });
  }, [setToken]);

  return { user, signIn, signOut, isAuthReady };
}
