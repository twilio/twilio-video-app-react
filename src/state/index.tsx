import React, { createContext, useContext, useReducer, useState } from 'react';
import { RecordingRules, RoomType } from '../types';
import { TwilioError } from 'twilio-video';
import { settingsReducer, initialSettings, Settings, SettingsAction } from './settings/settingsReducer';
import useActiveSinkId from './useActiveSinkId/useActiveSinkId';
import useFirebaseAuth from './useFirebaseAuth/useFirebaseAuth';
import usePasscodeAuth from './usePasscodeAuth/usePasscodeAuth';
import { User } from 'firebase';

export interface StateContextType {
  error: TwilioError | Error | null;
  setError(error: TwilioError | Error | null): void;
  getToken(name: string, room: string, passcode?: string): Promise<{ room_type: RoomType; token: string }>;
  user?: User | null | { displayName: undefined; photoURL: undefined; passcode?: string };
  signIn?(passcode?: string): Promise<void>;
  signOut?(): Promise<void>;
  isAuthReady?: boolean;
  isFetching: boolean;
  activeSinkId: string;
  setActiveSinkId(sinkId: string): void;
  settings: Settings;
  dispatchSetting: React.Dispatch<SettingsAction>;
  roomType?: RoomType;
  updateRecordingRules(room_sid: string, rules: RecordingRules): Promise<object>;
}

export const StateContext = createContext<StateContextType>(null!);

/*
  The 'react-hooks/rules-of-hooks' linting rules prevent React Hooks from being called
  inside of if() statements. This is because hooks must always be called in the same order
  every time a component is rendered. The 'react-hooks/rules-of-hooks' rule is disabled below
  because the "if (process.env.REACT_APP_SET_AUTH === 'firebase')" statements are evaluated
  at build time (not runtime). If the statement evaluates to false, then the code is not
  included in the bundle that is produced (due to tree-shaking). Thus, in this instance, it
  is ok to call hooks inside if() statements.
*/
export default function AppStateProvider(props: React.PropsWithChildren<{}>) {
  const [error, setError] = useState<TwilioError | null>(null);
  const [isFetching, setIsFetching] = useState(false);
  const [activeSinkId, setActiveSinkId] = useActiveSinkId();
  const [settings, dispatchSetting] = useReducer(settingsReducer, initialSettings);
  const [roomType, setRoomType] = useState<RoomType>();

  let contextValue = {
    error,
    setError,
    isFetching,
    activeSinkId,
    setActiveSinkId,
    settings,
    dispatchSetting,
    roomType,
  } as StateContextType;

  if (process.env.REACT_APP_SET_AUTH === 'firebase') {
    contextValue = {
      ...contextValue,
      ...useFirebaseAuth(), // eslint-disable-line react-hooks/rules-of-hooks
    };
  } else if (process.env.REACT_APP_SET_AUTH === 'passcode') {
    contextValue = {
      ...contextValue,
      ...usePasscodeAuth(), // eslint-disable-line react-hooks/rules-of-hooks
    };
  } else {
    contextValue = {
      ...contextValue,
      getToken: async (user_identity, room_name) => {
        const endpoint = process.env.REACT_APP_TOKEN_ENDPOINT || '/token';

        return fetch(endpoint, {
          method: 'POST',
          headers: {
            'content-type': 'application/json',
          },
          body: JSON.stringify({
            user_identity,
            room_name,
            create_conversation: process.env.REACT_APP_DISABLE_TWILIO_CONVERSATIONS !== 'true',
          }),
        }).then(res => res.json());
      },
      updateRecordingRules: async (room_sid, rules) => {
        const endpoint = process.env.REACT_APP_TOKEN_ENDPOINT || '/recordingrules';

        return fetch(endpoint, {
          headers: {
            'Content-Type': 'application/json',
          },
          body: JSON.stringify({ room_sid, rules }),
          method: 'POST',
        })
          .then(async res => {
            const jsonResponse = await res.json();

            if (!res.ok) {
              const recordingError = new Error(
                jsonResponse.error?.message || 'There was an error updating recording rules'
              );
              recordingError.code = jsonResponse.error?.code;
              return Promise.reject(recordingError);
            }

            return jsonResponse;
          })
          .catch(err => setError(err));
      },
    };
  }

  const getToken: StateContextType['getToken'] = (name, room) => {
    setIsFetching(true);
    return contextValue
      .getToken(name, room)
      .then(res => {
        setRoomType(res.room_type);
        setIsFetching(false);
        return res;
      })
      .catch(err => {
        setError(err);
        setIsFetching(false);
        return Promise.reject(err);
      });
  };

  const updateRecordingRules: StateContextType['updateRecordingRules'] = (room_sid, rules) => {
    setIsFetching(true);
    return contextValue
      .updateRecordingRules(room_sid, rules)
      .then(res => {
        setIsFetching(false);
        return res;
      })
      .catch(err => {
        setError(err);
        setIsFetching(false);
        return Promise.reject(err);
      });
  };

  return (
    <StateContext.Provider value={{ ...contextValue, getToken, updateRecordingRules }}>
      {props.children}
    </StateContext.Provider>
  );
}

export function useAppState() {
  const context = useContext(StateContext);
  if (!context) {
    throw new Error('useAppState must be used within the AppStateProvider');
  }
  return context;
}
