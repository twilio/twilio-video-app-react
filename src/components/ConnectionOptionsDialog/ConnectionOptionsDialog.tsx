import React, { useCallback } from 'react';
import {
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  DialogTitle,
  Divider,
  FormControl,
  Grid,
  InputLabel,
  MenuItem,
  Select,
  TextField,
  Theme,
  Typography,
} from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';
import { inputLabels, Settings } from '../../state/settings/settingsReducer';
import { useAppState } from '../../state';
import useRoomState from '../../hooks/useRoomState/useRoomState';

const useStyles = makeStyles((theme: Theme) => ({
  container: {
    width: '600px',
    minHeight: '400px',
    [theme.breakpoints.down('xs')]: {
      width: 'calc(100vw - 32px)',
    },
    '& .inputSelect': {
      width: 'calc(100% - 35px)',
    },
  },
  button: {
    float: 'right',
  },
  paper: {
    [theme.breakpoints.down('xs')]: {
      margin: '16px',
    },
  },
  formControl: {
    display: 'block',
    margin: '1.5em 0',
    '&:first-child': {
      margin: '0 0 1.5em 0',
    },
  },
  label: {
    width: '133%', // Labels have scale(0.75) applied to them, so this effectively makes the width 100%
  },
}));

const withDefault = (val?: string) => (typeof val === 'undefined' ? 'default' : val);

export default function ConnectionOptionsDialog({ open, onClose }: { open: boolean; onClose: () => void }) {
  const classes = useStyles();
  const { settings, dispatchSetting } = useAppState();
  const roomState = useRoomState();
  const isDisabled = roomState !== 'disconnected';

  const handleChange = useCallback(
    (e: React.ChangeEvent<{ value: unknown; name?: string }>) => {
      dispatchSetting({ name: e.target.name as keyof Settings, value: e.target.value as string });
    },
    [dispatchSetting]
  );

  const handleNumberChange = useCallback(
    (e: React.ChangeEvent<{ value: unknown; name?: string }>) => {
      if (!/[^\d]/.test(e.target.value as string)) handleChange(e);
    },
    [handleChange]
  );

  return (
    <Dialog open={open} onClose={onClose} classes={{ paper: classes.paper }}>
      <DialogTitle>Connection Settings</DialogTitle>
      <Divider />
      <DialogContent className={classes.container}>
        <Grid container spacing={2}>
          <Grid item xs={12}>
            <Typography hidden={!isDisabled} variant="body2">
              These settings cannot be changed when connected to a room.
            </Typography>
          </Grid>

          <Grid item sm={6} xs={12}>
            <FormControl className={classes.formControl}>
              <InputLabel id={inputLabels.dominantSpeakerPriority}>Dominant Speaker Priority:</InputLabel>
              <Select
                fullWidth
                disabled={isDisabled}
                name={inputLabels.dominantSpeakerPriority}
                label={inputLabels.dominantSpeakerPriority}
                value={withDefault(settings.dominantSpeakerPriority)}
                onChange={handleChange}
              >
                <MenuItem value="low">Low</MenuItem>
                <MenuItem value="standard">Standard</MenuItem>
                <MenuItem value="high">High</MenuItem>
                <MenuItem value="default">Server Default</MenuItem>
              </Select>
            </FormControl>

            <FormControl className={classes.formControl}>
              <InputLabel id={inputLabels.trackSwitchOffMode}>Track Switch Off Mode:</InputLabel>
              <Select
                fullWidth
                disabled={isDisabled}
                name={inputLabels.trackSwitchOffMode}
                label={inputLabels.trackSwitchOffMode}
                value={withDefault(settings.trackSwitchOffMode)}
                onChange={handleChange}
              >
                <MenuItem value="predicted">Predicted</MenuItem>
                <MenuItem value="detected">Detected</MenuItem>
                <MenuItem value="disabled">Disabled</MenuItem>
                <MenuItem value="default">Server Default</MenuItem>
              </Select>
            </FormControl>

            <FormControl className={classes.formControl}>
              <InputLabel id={inputLabels.bandwidthProfileMode}>Mode:</InputLabel>
              <Select
                fullWidth
                disabled={isDisabled}
                name={inputLabels.bandwidthProfileMode}
                label={inputLabels.bandwidthProfileMode}
                value={withDefault(settings.bandwidthProfileMode)}
                onChange={handleChange}
              >
                <MenuItem value="grid">Grid</MenuItem>
                <MenuItem value="collaboration">Collaboration</MenuItem>
                <MenuItem value="presentation">Presentation</MenuItem>
                <MenuItem value="default">Server Default</MenuItem>
              </Select>
            </FormControl>
          </Grid>
          <Grid item sm={6} xs={12}>
            <FormControl className={classes.formControl}>
              <InputLabel id={inputLabels.clientTrackSwitchOffControl}>Client Track Switch Off Control:</InputLabel>
              <Select
                fullWidth
                disabled={isDisabled}
                name={inputLabels.clientTrackSwitchOffControl}
                label={inputLabels.clientTrackSwitchOffControl}
                value={withDefault(settings.clientTrackSwitchOffControl)}
                onChange={handleChange}
              >
                <MenuItem value="auto">Auto</MenuItem>
                <MenuItem value="manual">Manual</MenuItem>
                <MenuItem value="default">Default</MenuItem>
              </Select>
            </FormControl>

            <FormControl className={classes.formControl}>
              <InputLabel id={inputLabels.contentPreferencesMode}>Content Preferences Mode:</InputLabel>
              <Select
                fullWidth
                disabled={isDisabled}
                name={inputLabels.contentPreferencesMode}
                label={inputLabels.contentPreferencesMode}
                value={withDefault(settings.contentPreferencesMode)}
                onChange={handleChange}
              >
                <MenuItem value="auto">Auto</MenuItem>
                <MenuItem value="manual">Manual</MenuItem>
                <MenuItem value="default">Default</MenuItem>
              </Select>
            </FormControl>

            <FormControl className={classes.formControl}>
              <TextField
                disabled={isDisabled}
                fullWidth
                id={inputLabels.maxAudioBitrate}
                label="Max Audio Bitrate"
                placeholder="Leave blank for no limit"
                name={inputLabels.maxAudioBitrate}
                value={withDefault(settings.maxAudioBitrate)}
                onChange={handleNumberChange}
              />
            </FormControl>
          </Grid>
        </Grid>
      </DialogContent>
      <Divider />
      <DialogActions>
        <Button className={classes.button} color="primary" variant="contained" onClick={onClose}>
          Done
        </Button>
      </DialogActions>
    </Dialog>
  );
}
