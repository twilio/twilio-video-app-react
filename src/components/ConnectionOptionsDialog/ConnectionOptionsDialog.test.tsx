import React from 'react';
import ConnectionOptionsDialog from './ConnectionOptionsDialog';
import { initialSettings } from '../../state/settings/settingsReducer';
import { Select, TextField } from '@material-ui/core';
import { shallow } from 'enzyme';
import { useAppState } from '../../state';
import useRoomState from '../../hooks/useRoomState/useRoomState';

jest.mock('../../hooks/useRoomState/useRoomState');
jest.mock('../../state');

const mockUseAppState = useAppState as jest.Mock<any>;
const mockUseRoomState = useRoomState as jest.Mock<any>;

const mockDispatchSetting = jest.fn();
mockUseAppState.mockImplementation(() => ({ settings: initialSettings, dispatchSetting: mockDispatchSetting }));

describe('the ConnectionOptionsDialog component', () => {
  afterEach(jest.clearAllMocks);

  describe('when not connected to a room', () => {
    mockUseRoomState.mockImplementation(() => 'disconnected');
    it('should render correctly', () => {
      const wrapper = shallow(<ConnectionOptionsDialog open={true} onClose={() => {}} />);
      expect(wrapper).toMatchSnapshot();
    });

    it('should dispatch settings changes', () => {
      const wrapper = shallow(<ConnectionOptionsDialog open={true} onClose={() => {}} />);
      wrapper
        .find(Select)
        .find({ name: 'dominantSpeakerPriority' })
        .simulate('change', { target: { value: 'testValue', name: 'dominantSpeakerPriority' } });
      expect(mockDispatchSetting).toHaveBeenCalledWith({ value: 'testValue', name: 'dominantSpeakerPriority' });
    });

    it('should not dispatch settings changes from a number field when there are non-digits in the value', () => {
      const wrapper = shallow(<ConnectionOptionsDialog open={true} onClose={() => {}} />);
      wrapper
        .find(TextField)
        .find({ name: 'maxAudioBitrate' })
        .simulate('change', { target: { value: '123456a', name: 'maxAudioBitrate' } });
      expect(mockDispatchSetting).not.toHaveBeenCalled();
    });

    it('should dispatch settings changes from a number field when there are only digits in the value', () => {
      const wrapper = shallow(<ConnectionOptionsDialog open={true} onClose={() => {}} />);
      wrapper
        .find(TextField)
        .find({ name: 'maxAudioBitrate' })
        .simulate('change', { target: { value: '123456', name: 'maxAudioBitrate' } });
      expect(mockDispatchSetting).toHaveBeenCalledWith({ value: '123456', name: 'maxAudioBitrate' });
    });
  });

  describe('when connected to a room', () => {
    mockUseRoomState.mockImplementation(() => 'connected');
    it('should render correctly', () => {
      const wrapper = shallow(<ConnectionOptionsDialog open={true} onClose={() => {}} />);
      expect(wrapper).toMatchSnapshot();
    });
  });
});
