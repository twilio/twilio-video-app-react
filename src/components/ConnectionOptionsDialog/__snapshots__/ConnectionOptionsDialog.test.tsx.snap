// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the ConnectionOptionsDialog component when connected to a room should render correctly 1`] = `
<WithStyles(ForwardRef(Dialog))
  classes={
    Object {
      "paper": "makeStyles-paper-3",
    }
  }
  onClose={[Function]}
  open={true}
>
  <WithStyles(ForwardRef(DialogTitle))>
    Connection Settings
  </WithStyles(ForwardRef(DialogTitle))>
  <WithStyles(ForwardRef(Divider)) />
  <WithStyles(ForwardRef(DialogContent))
    className="makeStyles-container-1"
  >
    <WithStyles(ForwardRef(Grid))
      container={true}
      spacing={2}
    >
      <WithStyles(ForwardRef(Grid))
        item={true}
        xs={12}
      >
        <WithStyles(ForwardRef(Typography))
          hidden={false}
          variant="body2"
        >
          These settings cannot be changed when connected to a room.
        </WithStyles(ForwardRef(Typography))>
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid))
        item={true}
        sm={6}
        xs={12}
      >
        <WithStyles(ForwardRef(FormControl))
          className="makeStyles-formControl-4"
        >
          <WithStyles(ForwardRef(InputLabel))
            id="dominantSpeakerPriority"
          >
            Dominant Speaker Priority:
          </WithStyles(ForwardRef(InputLabel))>
          <WithStyles(ForwardRef(Select))
            disabled={true}
            fullWidth={true}
            label="dominantSpeakerPriority"
            name="dominantSpeakerPriority"
            onChange={[Function]}
            value="standard"
          >
            <WithStyles(ForwardRef(MenuItem))
              value="low"
            >
              Low
            </WithStyles(ForwardRef(MenuItem))>
            <WithStyles(ForwardRef(MenuItem))
              value="standard"
            >
              Standard
            </WithStyles(ForwardRef(MenuItem))>
            <WithStyles(ForwardRef(MenuItem))
              value="high"
            >
              High
            </WithStyles(ForwardRef(MenuItem))>
            <WithStyles(ForwardRef(MenuItem))
              value="default"
            >
              Server Default
            </WithStyles(ForwardRef(MenuItem))>
          </WithStyles(ForwardRef(Select))>
        </WithStyles(ForwardRef(FormControl))>
        <WithStyles(ForwardRef(FormControl))
          className="makeStyles-formControl-4"
        >
          <WithStyles(ForwardRef(InputLabel))
            id="trackSwitchOffMode"
          >
            Track Switch Off Mode:
          </WithStyles(ForwardRef(InputLabel))>
          <WithStyles(ForwardRef(Select))
            disabled={true}
            fullWidth={true}
            label="trackSwitchOffMode"
            name="trackSwitchOffMode"
            onChange={[Function]}
            value="default"
          >
            <WithStyles(ForwardRef(MenuItem))
              value="predicted"
            >
              Predicted
            </WithStyles(ForwardRef(MenuItem))>
            <WithStyles(ForwardRef(MenuItem))
              value="detected"
            >
              Detected
            </WithStyles(ForwardRef(MenuItem))>
            <WithStyles(ForwardRef(MenuItem))
              value="disabled"
            >
              Disabled
            </WithStyles(ForwardRef(MenuItem))>
            <WithStyles(ForwardRef(MenuItem))
              value="default"
            >
              Server Default
            </WithStyles(ForwardRef(MenuItem))>
          </WithStyles(ForwardRef(Select))>
        </WithStyles(ForwardRef(FormControl))>
        <WithStyles(ForwardRef(FormControl))
          className="makeStyles-formControl-4"
        >
          <WithStyles(ForwardRef(InputLabel))
            id="bandwidthProfileMode"
          >
            Mode:
          </WithStyles(ForwardRef(InputLabel))>
          <WithStyles(ForwardRef(Select))
            disabled={true}
            fullWidth={true}
            label="bandwidthProfileMode"
            name="bandwidthProfileMode"
            onChange={[Function]}
            value="collaboration"
          >
            <WithStyles(ForwardRef(MenuItem))
              value="grid"
            >
              Grid
            </WithStyles(ForwardRef(MenuItem))>
            <WithStyles(ForwardRef(MenuItem))
              value="collaboration"
            >
              Collaboration
            </WithStyles(ForwardRef(MenuItem))>
            <WithStyles(ForwardRef(MenuItem))
              value="presentation"
            >
              Presentation
            </WithStyles(ForwardRef(MenuItem))>
            <WithStyles(ForwardRef(MenuItem))
              value="default"
            >
              Server Default
            </WithStyles(ForwardRef(MenuItem))>
          </WithStyles(ForwardRef(Select))>
        </WithStyles(ForwardRef(FormControl))>
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid))
        item={true}
        sm={6}
        xs={12}
      >
        <WithStyles(ForwardRef(FormControl))
          className="makeStyles-formControl-4"
        >
          <WithStyles(ForwardRef(InputLabel))
            id="clientTrackSwitchOffControl"
          >
            Client Track Switch Off Control:
          </WithStyles(ForwardRef(InputLabel))>
          <WithStyles(ForwardRef(Select))
            disabled={true}
            fullWidth={true}
            label="clientTrackSwitchOffControl"
            name="clientTrackSwitchOffControl"
            onChange={[Function]}
            value="auto"
          >
            <WithStyles(ForwardRef(MenuItem))
              value="auto"
            >
              Auto
            </WithStyles(ForwardRef(MenuItem))>
            <WithStyles(ForwardRef(MenuItem))
              value="manual"
            >
              Manual
            </WithStyles(ForwardRef(MenuItem))>
            <WithStyles(ForwardRef(MenuItem))
              value="default"
            >
              Default
            </WithStyles(ForwardRef(MenuItem))>
          </WithStyles(ForwardRef(Select))>
        </WithStyles(ForwardRef(FormControl))>
        <WithStyles(ForwardRef(FormControl))
          className="makeStyles-formControl-4"
        >
          <WithStyles(ForwardRef(InputLabel))
            id="contentPreferencesMode"
          >
            Content Preferences Mode:
          </WithStyles(ForwardRef(InputLabel))>
          <WithStyles(ForwardRef(Select))
            disabled={true}
            fullWidth={true}
            label="contentPreferencesMode"
            name="contentPreferencesMode"
            onChange={[Function]}
            value="auto"
          >
            <WithStyles(ForwardRef(MenuItem))
              value="auto"
            >
              Auto
            </WithStyles(ForwardRef(MenuItem))>
            <WithStyles(ForwardRef(MenuItem))
              value="manual"
            >
              Manual
            </WithStyles(ForwardRef(MenuItem))>
            <WithStyles(ForwardRef(MenuItem))
              value="default"
            >
              Default
            </WithStyles(ForwardRef(MenuItem))>
          </WithStyles(ForwardRef(Select))>
        </WithStyles(ForwardRef(FormControl))>
        <WithStyles(ForwardRef(FormControl))
          className="makeStyles-formControl-4"
        >
          <WithStyles(ForwardRef(TextField))
            disabled={true}
            fullWidth={true}
            id="maxAudioBitrate"
            label="Max Audio Bitrate"
            name="maxAudioBitrate"
            onChange={[Function]}
            placeholder="Leave blank for no limit"
            value="16000"
          />
        </WithStyles(ForwardRef(FormControl))>
      </WithStyles(ForwardRef(Grid))>
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(DialogContent))>
  <WithStyles(ForwardRef(Divider)) />
  <WithStyles(ForwardRef(DialogActions))>
    <WithStyles(ForwardRef(Button))
      className="makeStyles-button-2"
      color="primary"
      onClick={[Function]}
      variant="contained"
    >
      Done
    </WithStyles(ForwardRef(Button))>
  </WithStyles(ForwardRef(DialogActions))>
</WithStyles(ForwardRef(Dialog))>
`;

exports[`the ConnectionOptionsDialog component when not connected to a room should render correctly 1`] = `
<WithStyles(ForwardRef(Dialog))
  classes={
    Object {
      "paper": "makeStyles-paper-3",
    }
  }
  onClose={[Function]}
  open={true}
>
  <WithStyles(ForwardRef(DialogTitle))>
    Connection Settings
  </WithStyles(ForwardRef(DialogTitle))>
  <WithStyles(ForwardRef(Divider)) />
  <WithStyles(ForwardRef(DialogContent))
    className="makeStyles-container-1"
  >
    <WithStyles(ForwardRef(Grid))
      container={true}
      spacing={2}
    >
      <WithStyles(ForwardRef(Grid))
        item={true}
        xs={12}
      >
        <WithStyles(ForwardRef(Typography))
          hidden={false}
          variant="body2"
        >
          These settings cannot be changed when connected to a room.
        </WithStyles(ForwardRef(Typography))>
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid))
        item={true}
        sm={6}
        xs={12}
      >
        <WithStyles(ForwardRef(FormControl))
          className="makeStyles-formControl-4"
        >
          <WithStyles(ForwardRef(InputLabel))
            id="dominantSpeakerPriority"
          >
            Dominant Speaker Priority:
          </WithStyles(ForwardRef(InputLabel))>
          <WithStyles(ForwardRef(Select))
            disabled={true}
            fullWidth={true}
            label="dominantSpeakerPriority"
            name="dominantSpeakerPriority"
            onChange={[Function]}
            value="standard"
          >
            <WithStyles(ForwardRef(MenuItem))
              value="low"
            >
              Low
            </WithStyles(ForwardRef(MenuItem))>
            <WithStyles(ForwardRef(MenuItem))
              value="standard"
            >
              Standard
            </WithStyles(ForwardRef(MenuItem))>
            <WithStyles(ForwardRef(MenuItem))
              value="high"
            >
              High
            </WithStyles(ForwardRef(MenuItem))>
            <WithStyles(ForwardRef(MenuItem))
              value="default"
            >
              Server Default
            </WithStyles(ForwardRef(MenuItem))>
          </WithStyles(ForwardRef(Select))>
        </WithStyles(ForwardRef(FormControl))>
        <WithStyles(ForwardRef(FormControl))
          className="makeStyles-formControl-4"
        >
          <WithStyles(ForwardRef(InputLabel))
            id="trackSwitchOffMode"
          >
            Track Switch Off Mode:
          </WithStyles(ForwardRef(InputLabel))>
          <WithStyles(ForwardRef(Select))
            disabled={true}
            fullWidth={true}
            label="trackSwitchOffMode"
            name="trackSwitchOffMode"
            onChange={[Function]}
            value="default"
          >
            <WithStyles(ForwardRef(MenuItem))
              value="predicted"
            >
              Predicted
            </WithStyles(ForwardRef(MenuItem))>
            <WithStyles(ForwardRef(MenuItem))
              value="detected"
            >
              Detected
            </WithStyles(ForwardRef(MenuItem))>
            <WithStyles(ForwardRef(MenuItem))
              value="disabled"
            >
              Disabled
            </WithStyles(ForwardRef(MenuItem))>
            <WithStyles(ForwardRef(MenuItem))
              value="default"
            >
              Server Default
            </WithStyles(ForwardRef(MenuItem))>
          </WithStyles(ForwardRef(Select))>
        </WithStyles(ForwardRef(FormControl))>
        <WithStyles(ForwardRef(FormControl))
          className="makeStyles-formControl-4"
        >
          <WithStyles(ForwardRef(InputLabel))
            id="bandwidthProfileMode"
          >
            Mode:
          </WithStyles(ForwardRef(InputLabel))>
          <WithStyles(ForwardRef(Select))
            disabled={true}
            fullWidth={true}
            label="bandwidthProfileMode"
            name="bandwidthProfileMode"
            onChange={[Function]}
            value="collaboration"
          >
            <WithStyles(ForwardRef(MenuItem))
              value="grid"
            >
              Grid
            </WithStyles(ForwardRef(MenuItem))>
            <WithStyles(ForwardRef(MenuItem))
              value="collaboration"
            >
              Collaboration
            </WithStyles(ForwardRef(MenuItem))>
            <WithStyles(ForwardRef(MenuItem))
              value="presentation"
            >
              Presentation
            </WithStyles(ForwardRef(MenuItem))>
            <WithStyles(ForwardRef(MenuItem))
              value="default"
            >
              Server Default
            </WithStyles(ForwardRef(MenuItem))>
          </WithStyles(ForwardRef(Select))>
        </WithStyles(ForwardRef(FormControl))>
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid))
        item={true}
        sm={6}
        xs={12}
      >
        <WithStyles(ForwardRef(FormControl))
          className="makeStyles-formControl-4"
        >
          <WithStyles(ForwardRef(InputLabel))
            id="clientTrackSwitchOffControl"
          >
            Client Track Switch Off Control:
          </WithStyles(ForwardRef(InputLabel))>
          <WithStyles(ForwardRef(Select))
            disabled={true}
            fullWidth={true}
            label="clientTrackSwitchOffControl"
            name="clientTrackSwitchOffControl"
            onChange={[Function]}
            value="auto"
          >
            <WithStyles(ForwardRef(MenuItem))
              value="auto"
            >
              Auto
            </WithStyles(ForwardRef(MenuItem))>
            <WithStyles(ForwardRef(MenuItem))
              value="manual"
            >
              Manual
            </WithStyles(ForwardRef(MenuItem))>
            <WithStyles(ForwardRef(MenuItem))
              value="default"
            >
              Default
            </WithStyles(ForwardRef(MenuItem))>
          </WithStyles(ForwardRef(Select))>
        </WithStyles(ForwardRef(FormControl))>
        <WithStyles(ForwardRef(FormControl))
          className="makeStyles-formControl-4"
        >
          <WithStyles(ForwardRef(InputLabel))
            id="contentPreferencesMode"
          >
            Content Preferences Mode:
          </WithStyles(ForwardRef(InputLabel))>
          <WithStyles(ForwardRef(Select))
            disabled={true}
            fullWidth={true}
            label="contentPreferencesMode"
            name="contentPreferencesMode"
            onChange={[Function]}
            value="auto"
          >
            <WithStyles(ForwardRef(MenuItem))
              value="auto"
            >
              Auto
            </WithStyles(ForwardRef(MenuItem))>
            <WithStyles(ForwardRef(MenuItem))
              value="manual"
            >
              Manual
            </WithStyles(ForwardRef(MenuItem))>
            <WithStyles(ForwardRef(MenuItem))
              value="default"
            >
              Default
            </WithStyles(ForwardRef(MenuItem))>
          </WithStyles(ForwardRef(Select))>
        </WithStyles(ForwardRef(FormControl))>
        <WithStyles(ForwardRef(FormControl))
          className="makeStyles-formControl-4"
        >
          <WithStyles(ForwardRef(TextField))
            disabled={true}
            fullWidth={true}
            id="maxAudioBitrate"
            label="Max Audio Bitrate"
            name="maxAudioBitrate"
            onChange={[Function]}
            placeholder="Leave blank for no limit"
            value="16000"
          />
        </WithStyles(ForwardRef(FormControl))>
      </WithStyles(ForwardRef(Grid))>
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(DialogContent))>
  <WithStyles(ForwardRef(Divider)) />
  <WithStyles(ForwardRef(DialogActions))>
    <WithStyles(ForwardRef(Button))
      className="makeStyles-button-2"
      color="primary"
      onClick={[Function]}
      variant="contained"
    >
      Done
    </WithStyles(ForwardRef(Button))>
  </WithStyles(ForwardRef(DialogActions))>
</WithStyles(ForwardRef(Dialog))>
`;
